### PR TITLE
implementation EGP for RPC

### PR DESCRIPTION
**Pull request by joanestebanr**
Implementation of Estimated GasPrice logic for RPC
https://app.diagrams.net/#G1rb3FKaEg4fnmu6mMdvvr-h36FjzEY4ek

### CODE REVIEW

File: pool/pool.go [hunk 1]
```diff
@@ -420,6 +468,9 @@ func (p *Pool) validateTx(ctx context.Context, poolTx Transaction) error {
 	// Reject transactions with a gas price lower than the minimum gas price
 	p.minSuggestedGasPriceMux.RLock()
 	gasPriceCmp := poolTx.GasPrice().Cmp(p.minSuggestedGasPrice)
+	if gasPriceCmp == -1 {
+		log.Debugf("low gas price: minSuggestedGasPrice %v got %v", p.minSuggestedGasPrice, poolTx.GasPrice())
```
Comment (tclemos) [inline pool/pool.go:472]: can we group this `if block` with `if block` below?

**Review by tclemos — APPROVED** [review]

**Review by Psykepro — APPROVED** [review]

File: pool/pool.go [hunk 2]
```diff
@@ -223,12 +228,57 @@ func (p *Pool) StoreTx(ctx context.Context, tx types.Transaction, ip string, isW
 		}
 	}
 
+	gasPrices, err := p.GetGasPrices(ctx)
```
Comment (agnusmor) [inline pool/pool.go]: This makes a query to the database to get the gasPrices, this could have some impact in the performance, as we are quering for each txs we need to validate. We have the pool.GetL1AndL2GasPrice function that gets this info from a shared struct that is updated each 5 seconds (IntervalToRefreshGasPrices), it is update in the refreshGasPrices function (that is called from a go function). The pool.GetL1AndL2GasPrice is the function used in the sequencer to get the proces used during the egp calculation

File: pool/pool.go [hunk 3]
```diff
@@ -223,12 +228,57 @@ func (p *Pool) StoreTx(ctx context.Context, tx types.Transaction, ip string, isW
...
+	if err != nil {
+		log.Error("failed to load gas prices")
+	}
+	err = p.ValidateEffectiveGasPrice(ctx, tx, preExecutionResponse.txResponse.GasUsed, gasPrices)
```
Comment (agnusmor) [inline pool/pool.go]: I think this function must be calle `ValidateBreakEvenGasPrice` as really this function is checking the breakEvenGasPrice, the EffectiveGasPrice is a different calculation/value that takes also into account the ratio between tx.GasPrice and L2GasPrice

File: pool/pool.go [hunk 4]
```diff
@@ -223,12 +228,57 @@ func (p *Pool) StoreTx(ctx context.Context, tx types.Transaction, ip string, isW
...
+func (p *Pool) ValidateEffectiveGasPrice(ctx context.Context, tx types.Transaction, preExecutionGasUsed uint64, gasPrices GasPrices) error {
+	breakEvenGasPrice, err := p.effectiveGasPrice.CalculateBreakEvenGasPrice(tx.Data(), tx.GasPrice(), preExecutionGasUsed, gasPrices.L1GasPrice)
+	if err != nil {
+		log.Errorf("error calculating break even gas price: %v", err)
+		return err
+	}
+	margin := big.NewInt(0).SetUint64(p.cfg.EffectiveGasPrice.MarginBreakEven)
+	breakEvenGasPricePercentApplied := big.NewInt(0).Mul(breakEvenGasPrice, margin)
+	breakEvenGasPricePercentApplied = breakEvenGasPricePercentApplied.Div(breakEvenGasPricePercentApplied, big.NewInt(percentFactor))
+	log.Infof("EGP breakEvenGasPrice: tx:%s breakEvenGasPrice:%v margin:%d breakEvenGasPricePercentApplied:%v L1GasPrice:%d L2GasPrice:%d tx.GasPrice():%v EGPenabled:%t",
```
Comment (agnusmor) [inline pool/pool.go]: In the sequencer side I start this log with "egp-log: " ("egp-log: breakEvenGasPrice: .....") maybe we can use the same format to be consistent, anyway we can review is "egp-log" is ok or use other prefix

File: pool/pool.go [hunk 5]
```diff
@@ -223,12 +228,57 @@ func (p *Pool) StoreTx(ctx context.Context, tx types.Transaction, ip string, isW
...
+				return ErrEffectiveGasPriceGasPriceTooLow
+			}
+			log.Debug("EGP tx:%s Accepted tx with gasPrice lower than L2GasPrice but EGP is disabled", tx.Hash().String())
+		} else {
+			// accept
+			log.Debugf("EGP tx:%s Accepted tx with loss because gasPrice in tx (%v) >= current L2GasPrice (%v)", tx.Hash().String(), tx.GasPrice(), L2GasPrice)
+		}
+		loss := big.NewInt(0).Set(breakEvenGasPrice)
+		loss = loss.Sub(loss, tx.GasPrice())
+		log.Warnf("EGP tx:%s Accepted tx with loss of  %s (breakEven=%s - txGas=%s)", tx.Hash().String(), loss.String(), breakEvenGasPrice.String(), tx.GasPrice().String())
```
Comment (agnusmor) [inline pool/pool.go]: In this case, in the sequencer, I am using prefix "egp-loss:"

**Review by agnusmor — APPROVED** [review]

File: pool/pool.go [hunk 6]
```diff
@@ -222,13 +227,56 @@ func (p *Pool) StoreTx(ctx context.Context, tx types.Transaction, ip string, isW
...
+		// check against gasPrice now
+		L2GasPrice := big.NewInt(0).SetUint64(gasPrices.L2GasPrice)
+		if tx.GasPrice().Cmp(L2GasPrice) == -1 { // tx.GasPrice() < gasPrices.L2GasPrice
+			// Reject transaction if EffetiveGasPrice is enabled
+			if p.cfg.EffectiveGasPrice.Enabled {
+				// reject
+				log.Debugf("egp-log:  tx:%s Reject tx with gasPrice lower than L2GasPrice", tx.Hash().String())
+				return ErrEffectiveGasPriceGasPriceTooLow
+			}
+			log.Debug("egp-log:  tx:%s Accepted tx with gasPrice lower than L2GasPrice but EGP is disabled", tx.Hash().String())
```
Comment (agnusmor) [inline pool/pool.go]: 2 whites spaces after "egp-log:"

File: pool/pool.go [hunk 7]
```diff
@@ -222,13 +227,56 @@ func (p *Pool) StoreTx(ctx context.Context, tx types.Transaction, ip string, isW
...
+		p.cfg.EffectiveGasPrice.MarginBreakEven, breakEvenGasPricePercentApplied, gasPrices.L1GasPrice, gasPrices.L2GasPrice, tx.GasPrice(),
+		p.cfg.EffectiveGasPrice.Enabled)
+	if breakEvenGasPricePercentApplied.Cmp(tx.GasPrice()) == 1 { // breakEvenGasPricePercentApplied > tx.GasPrice()
+		// check against gasPrice now
+		L2GasPrice := big.NewInt(0).SetUint64(gasPrices.L2GasPrice)
+		if tx.GasPrice().Cmp(L2GasPrice) == -1 { // tx.GasPrice() < gasPrices.L2GasPrice
+			// Reject transaction if EffetiveGasPrice is enabled
+			if p.cfg.EffectiveGasPrice.Enabled {
+				// reject
+				log.Debugf("egp-log:  tx:%s Reject tx with gasPrice lower than L2GasPrice", tx.Hash().String())
```
Comment (agnusmor) [inline pool/pool.go]: 2 whites spaces after "egp-log:"

File: pool/pool.go [hunk 8]
```diff
@@ -222,13 +227,56 @@ func (p *Pool) StoreTx(ctx context.Context, tx types.Transaction, ip string, isW
...
+				return ErrEffectiveGasPriceGasPriceTooLow
+			}
+			log.Debug("egp-log:  tx:%s Accepted tx with gasPrice lower than L2GasPrice but EGP is disabled", tx.Hash().String())
+		} else {
+			// accept
+			log.Debugf("egp-log: egp-loss: tx:%s Accepted tx with loss because gasPrice in tx (%v) >= current L2GasPrice (%v)", tx.Hash().String(), tx.GasPrice(), L2GasPrice)
```
Comment (agnusmor) [inline pool/pool.go]: I think this should be "egp-log" and log only one "egp-loss" (line 275). The idea is to have only one "egp-loss" for each tx since in this way it we will be more easy when filering logs looking for txs where we loss "money"

File: pool/pool.go [hunk 9]
```diff
@@ -222,13 +227,56 @@ func (p *Pool) StoreTx(ctx context.Context, tx types.Transaction, ip string, isW
...
+		}
+		loss := big.NewInt(0).Set(breakEvenGasPrice)
+		loss = loss.Sub(loss, tx.GasPrice())
+		log.Warnf("egp-log: egp-loss:  tx:%s Accepted tx with loss of  %s (breakEven=%s - txGas=%s)", tx.Hash().String(), loss.String(), breakEvenGasPrice.String(), tx.GasPrice().String())
```
Comment (agnusmor) [inline pool/pool.go]: Remove "egp-log:", the log should start with "egp-loss:" (same behaviour as sequencer)
